Update documentation to reflect breaking changes

bguyza opened the pull request:
This PR mostly addresses breaking changes related to the `User` domain object changing for MFA.  It also addresses breaking changes to the Integrations API. 

-  User domain object
-  Integrations API, removed Twilio
-  Remove `user.twoFactorEnabled` field from custom forms
-  ~Remove twilio tutorial (Comment from Dan M, I don't think we should do this for anyone who is pre 1.26 for a few months).~

mooreds commented on site/docs/v1/tech/apis/_import-users-request-body.adoc:
```diff
@@ -163,7 +163,45 @@ When an API key is used that is locked to a Tenant or the tenant `X-FusionAuth-T
 
-[field]#users``[x]``.twoFactorDelivery# [type]#[String]# [optional]#Optional# [default]#defaults to `None`#::
+[field]#users``[x]``.twoFactor.methods``[x]``.authenticator.algorithm# [type]#[TOTPAlgorithm]# [optional]#Optional#::
+The algorithm used by the TOTP authenticator. With the current implementation, this will always be `HmacSHA1`.
```
"The algorithm used by the TOTP authenticator. With the current implementation, this will always be `HmacSHA1`." =>  "The algorithm used by the TOTP authenticator. With the current implementation, this must be `HmacSHA1` if provided. Any other value will be ignored."

Since this is a request, correct?

bguyza replied:
Good call, that is correct.

mooreds commented on site/docs/v1/tech/apis/_import-users-request-body.adoc:
```diff
@@ -163,7 +163,45 @@ When an API key is used that is locked to a Tenant or the tenant `X-FusionAuth-T
+The algorithm used by the TOTP authenticator. With the current implementation, this will always be `HmacSHA1`.
+
+[field]#users``[x]``.twoFactor.methods``[x]``.authenticator.codeLength# [type]#[Integer]# [optional]#Optional#::
+The length of code generated by the TOTP. With the current implementation, this will always be `6`.
```
same as above comment.

mooreds commented on site/docs/v1/tech/apis/_import-users-request-body.adoc:
```diff
@@ -163,7 +163,45 @@ When an API key is used that is locked to a Tenant or the tenant `X-FusionAuth-T
+The length of code generated by the TOTP. With the current implementation, this will always be `6`.
+
+[field]#users``[x]``.twoFactor.methods``[x]``.authenticator.timeStep# [type]#[Integer]# [optional]#Optional#::
+The time-step size in seconds. With the current implementation, this will always be 30.
```
same as the comment about hmacsha1

mooreds commented on site/docs/v1/tech/apis/_import-users-request-body.adoc:
```diff
@@ -163,7 +163,45 @@ When an API key is used that is locked to a Tenant or the tenant `X-FusionAuth-T
+[field]#users``[x]``.twoFactor.methods``[x]``.email# [type]#[String]# [optional]#Optional#::
+The value of the email address for this method. Only present if [field]#user.twoFactor.methods``[x]``.method# is `email`.
+
+[field]#users``[x]``.twoFactor.methods``[x]``.id# [type]#[UUID]# [optional]#Optional#::
```
This is not a UUID. This is a string. Either that or `methodId` is incorrect here: https://fusionauth.io/docs/v1/tech/apis/two-factor/#disable-multi-factor

bguyza replied:
Yep yep, its a String

mooreds commented on site/docs/v1/tech/apis/_family-pending-response-body.adoc:
```diff
@@ -40,16 +40,58 @@ The Id of the Tenant that this User belongs to.
+[field]#users``[x]``.twoFactor.methods``[x]``.email# [type]#[String]#::
+The value of the email address for this method. Only present if [field]#user.twoFactor.methods``[x]``.method# is `email`.
+
+[field]#users``[x]``.twoFactor.methods``[x]``.id# [type]#[UUID]#::
```
This is not a UUID. This is a string. Either that or `methodId` is incorrect here: https://fusionauth.io/docs/v1/tech/apis/two-factor/#disable-multi-factor

mooreds commented on site/docs/v1/tech/apis/_integration-request-body.adoc:
```diff
@@ -25,22 +25,32 @@ True if the Kafka integration is enabled.
-[field]#integration.twilio.accountSID# [type]#[String]# [optional]#Optional#::
-The Twilio Account ID to use when connecting to the Twilio API. This can be found in your Twilio dashboard. Required when the Twilio integration is enabled.
+[field]#integration.twilio.accountSID# [type]#[String]# [optional]#Optional# [deprecated]#Deprecated#::
+blah blah blahThe Twilio Account ID to use when connecting to the Twilio API. This can be found in your Twilio dashboard. Required when the Twilio integration is enabled.
```
I don't think you want 'blah blah blah' here. :)

bguyza replied:
Oops! This was when I was getting frustrated when the page wasn't re-compiling.  Hence the slack convo where I learned I need to touch the file that includes this one! Funny...

mooreds commented on site/docs/v1/tech/apis/_user-registration-combined-request-body.adoc:
```diff
@@ -109,7 +109,45 @@ The User's preferred timezone. The string must be in an https://www.iana.org/tim
 
-[field]#users.twoFactorDelivery# [type]#[String]# [optional]#Optional# [default]#defaults to `None`#::
+[field]#user.twoFactor.methods``[x]``.authenticator.algorithm# [type]#[TOTPAlgorithm]# [optional]#Optional#::
+The algorithm used by the TOTP authenticator. With the current implementation, this will always be `HmacSHA1`.
```
See comments "The algorithm used by the TOTP authenticator. With the current implementation, this will always be `HmacSHA1`." =>  "The algorithm used by the TOTP authenticator. With the current implementation, this must be `HmacSHA1` if provided. Any other value will be ignored."

Since this is a request, correct?

mooreds commented on site/docs/v1/tech/apis/_user-registration-combined-request-body.adoc:
```diff
@@ -109,7 +109,45 @@ The User's preferred timezone. The string must be in an https://www.iana.org/tim
+The algorithm used by the TOTP authenticator. With the current implementation, this will always be `HmacSHA1`.
+
+[field]#user.twoFactor.methods``[x]``.authenticator.codeLength# [type]#[Integer]# [optional]#Optional#::
+The length of code generated by the TOTP. With the current implementation, this will always be `6`.
```
See comment above

mooreds commented on site/docs/v1/tech/apis/_user-registration-combined-request-body.adoc:
```diff
@@ -109,7 +109,45 @@ The User's preferred timezone. The string must be in an https://www.iana.org/tim
+The length of code generated by the TOTP. With the current implementation, this will always be `6`.
+
+[field]#user.twoFactor.methods``[x]``.authenticator.timeStep# [type]#[Integer]# [optional]#Optional#::
+The time-step size in seconds. With the current implementation, this will always be 30.
```
See hmacsha comment above.

mooreds commented on site/docs/v1/tech/apis/_user-registration-combined-request-body.adoc:
```diff
@@ -109,7 +109,45 @@ The User's preferred timezone. The string must be in an https://www.iana.org/tim
+[field]#user.twoFactor.methods``[x]``.email# [type]#[String]# [optional]#Optional#::
+The value of the email address for this method. Only present if [field]#user.twoFactor.methods``[x]``.method# is `email`.
+
+[field]#user.twoFactor.methods``[x]``.id# [type]#[UUID]# [optional]#Optional#::
```
Not a UUID (as per other `methods` comments) but a string.

mooreds replied:
Also can you even provide the id?

bguyza replied:
Good catch - you should not be passing an Id here.

mooreds commented on site/docs/v1/tech/apis/_user-registration-combined-request-body.adoc:
```diff
@@ -109,7 +109,45 @@ The User's preferred timezone. The string must be in an https://www.iana.org/tim
+[field]#user.twoFactor.methods``[x]``.id# [type]#[UUID]# [optional]#Optional#::
+The unique Id of the method.
+
+[field]#user.twoFactor.methods``[x]``.lastUsed# [type]#[Boolean]# [optional]#Optional#::
```
This is something that can be set?

bguyza replied:
Nope - copy/pasta :(

mooreds commented on site/docs/v1/tech/apis/_user-registration-combined-response-body.adoc:
```diff
@@ -103,16 +103,51 @@ An array of locale strings that give, in order, the User's preferred languages.
+[field]#user.twoFactor.methods``[x]``.email# [type]#[String]#::
+The value of the email address for this method. Only present if [field]#user.twoFactor.methods``[x]``.method# is `email`.
+
+[field]#user.twoFactor.methods``[x]``.id# [type]#[UUID]#::
```
not a UUID

mooreds commented on site/docs/v1/tech/apis/_user-response-body.adoc:
```diff
@@ -174,16 +174,51 @@ The Id of the Tenant that this User belongs to.
 The User's preferred timezone. This can be used as a default to display instants, and it is recommended that you allow User's to change this per-session. The string will be in an https://www.iana.org/time-zones[IANA] time zone format.
 
-[field]#user.twoFactorDelivery# [type]#[String]#::
+[field]#user.twoFactor.methods``[x]``.authenticator.algorithm# [type]#[TOTPAlgorithm]#::
```
What is that type? TOTPAlgorithm? Isn't it a string?

bguyza replied:
Ah yes. This is the enum name.  It should be String.  Thank you for calling this out.

mooreds commented on site/docs/v1/tech/apis/_users-response-body.adoc:
```diff
@@ -143,16 +143,53 @@ The Id of the Tenant that this User belongs to.
 The User's preferred timezone. This can be used as a default to display instants, and it is recommended that you allow User's to change this per-session. The string will be in an https://www.iana.org/time-zones[IANA] time zone format.
 
-[field]#users``[x]``.twoFactorDelivery# [type]#[String]#::
+[field]#users``[x]``.twoFactor.methods``[x]``.authenticator.algorithm# [type]#[TOTPAlgorithm]#::
```
What is that type? TOTPAlgorithm? Isn't it a string?

mooreds commented on site/docs/v1/tech/apis/_family-pending-response-body.adoc:
```diff
@@ -40,16 +40,58 @@ The Id of the Tenant that this User belongs to.
 The User's preferred timezone. This can be used as a default to display instants, and it is recommended that you allow User's to change this per-session. The string will be in an https://www.iana.org/time-zones[IANA] time zone format.
 
-[field]#users``[x]``.twoFactorDelivery# [type]#[String]#::
+[field]#users``[x]``.twoFactor.methods``[x]``.authenticator.algorithm# [type]#[TOTPAlgorithm]#::
```
What is that type? TOTPAlgorithm? Isn't it a string?

mooreds commented on site/docs/v1/tech/apis/_import-users-request-body.adoc:
```diff
@@ -163,7 +163,45 @@ When an API key is used that is locked to a Tenant or the tenant `X-FusionAuth-T
 The User's preferred timezone. The string must be in an https://www.iana.org/time-zones[IANA] time zone format.
 
-[field]#users``[x]``.twoFactorDelivery# [type]#[String]# [optional]#Optional# [default]#defaults to `None`#::
+[field]#users``[x]``.twoFactor.methods``[x]``.authenticator.algorithm# [type]#[TOTPAlgorithm]# [optional]#Optional#::
```
What is that type? TOTPAlgorithm? Isn't it a string?

mooreds commented on site/docs/v1/tech/apis/_user-registration-combined-request-body.adoc:
```diff
@@ -109,7 +109,45 @@ The User's preferred timezone. The string must be in an https://www.iana.org/tim
 `America/Denver` or `US/Mountain`
 
-[field]#users.twoFactorDelivery# [type]#[String]# [optional]#Optional# [default]#defaults to `None`#::
+[field]#user.twoFactor.methods``[x]``.authenticator.algorithm# [type]#[TOTPAlgorithm]# [optional]#Optional#::
```
What is that type? TOTPAlgorithm? Isn't it a string?

mooreds commented on site/docs/v1/tech/apis/_user-registration-combined-response-body.adoc:
```diff
@@ -103,16 +103,51 @@ An array of locale strings that give, in order, the User's preferred languages.
 The User's preferred timezone. The string will be in an https://www.iana.org/time-zones[IANA] time zone format.
 
-[field]#users.twoFactorDelivery# [type]#[String]#::
+[field]#user.twoFactor.methods``[x]``.authenticator.algorithm# [type]#[TOTPAlgorithm]#::
```
What is that type? TOTPAlgorithm? Isn't it a string?

mooreds commented on site/docs/v1/tech/apis/_user-request-body.adoc:
```diff
@@ -114,7 +114,45 @@ The User's preferred timezone. The string must be in an https://www.iana.org/tim
 `America/Denver` or `US/Mountain`
 
-[field]#users.twoFactorDelivery# [type]#[String]# [optional]#Optional# [default]#defaults to `None`#::
+[field]#user.twoFactor.methods``[x]``.authenticator.algorithm# [type]#[TOTPAlgorithm]# [optional]#Optional#::
```
What is that type? TOTPAlgorithm? Isn't it a string?

mooreds commented on site/docs/v1/tech/apis/_user-response-body.adoc:
```diff
@@ -174,16 +174,51 @@ The Id of the Tenant that this User belongs to.
+[field]#user.twoFactor.methods``[x]``.email# [type]#[String]#::
+The value of the email address for this method. Only present if [field]#user.twoFactor.methods``[x]``.method# is `email`.
+
+[field]#user.twoFactor.methods``[x]``.id# [type]#[UUID]#::
```
Not a UUID.

mooreds commented on site/docs/v1/tech/apis/_users-response-body.adoc:
```diff
@@ -143,16 +143,53 @@ The Id of the Tenant that this User belongs to.
+[field]#users``[x]``.twoFactor.methods``[x]``.email# [type]#[String]#::
+The value of the email address for this method. Only present if [field]#user.twoFactor.methods``[x]``.method# is `email`.
+
+[field]#users``[x]``.twoFactor.methods``[x]``.id# [type]#[UUID]#::
```
Not a UUID.

mooreds commented on site/docs/v1/tech/apis/users.adoc:
```diff
@@ -492,7 +492,7 @@ include::docs/v1/tech/apis/_standard-post-response-codes.adoc[]
 ====
 
-THis API is used to import refresh tokens from an external system, this would generally be done during an initial user import or as an auxiliary step during a migration strategy.
+This API is used to import refresh tokens from an external system, this would generally be done during an initial user import or as an auxiliary step during a migration strategy.
```
/docs/v1/tech/integrations/twilio/ needs to be removed or marked deprecated, too.

bguyza replied:
It looks like this is being handled in this PR: https://github.com/FusionAuth/fusionauth-site/pull/590 so I'm going to hold off.

mooreds reviewed (changes requested):
Good first pass. I noted a few things that I think need to change. 

biggest one was that the Twilio integrations page should be removed or marked deprecated.

Most of the other ones were small, repetitive changes.

bguyza commented:
@mooreds 
Thanks for the review.  This should be ready for a second look.

 In summary:
- I left the Twilio integrations page as-is since it looks like Josh has handled that in PR #590 
- Updated `twoFactor.methods.id` to be of type `String` and not `UUID`
- Removed `twoFactor.methods.id` and `twoFactor.methods.lastUsed` from all requests. They shouldn't be there.
- Updated verbiage for `authenticator` attribute descriptions as requested.
- Replaced `TOTPAlgorithm` with `String` (Not sure why I put the enum name in there in the first place).  

Also, I usually mark each conversation as resolved as I fix them almost akin to a checklist, but is it more helpful to leave them out there so its easier for you to see when following up?  I left them unresolved to get your feedback. 

Thanks.

mooreds commented:
LGTM. 

I prefer if conversations are marked resolved, that way I know you've addressed any issues.
